Client/Frontend/Shields/ShieldsViewController.swift

```diff
@@ -0,0 +1,164 @@
+    
+    init(url: URL, shieldBlockStats: ShieldBlockedStats?) {
+        self.url = url
+        self.shieldBlockStats = shieldBlockStats
+        
+        super.init(nibName: nil, bundle: nil)
+        
+        shieldsView.shieldsContainerStackView.hostLabel.text = url.normalizedHost
+        
+        updateToggleStatus()
+        updateShieldBlockStats()
+    }
+    
+    // MARK: - State
+    
+    private func updateToggleStatus() {
+        shieldControlMapping.forEach { shield, view, option in
+            if let option = option {
+                // Sets the default setting
+                view.toggleSwitch.isOn = option.value
+            } else {
+                // The only "shield" that doesn't actually have a global preference is the option to disable all shields
+                // Therefore its the only shield we have to give a default value of true
+                view.toggleSwitch.isOn = shield == .AllOff
+            }
+            if let host = url.normalizedHost, let shieldState = BraveShieldState.getStateForDomain(host) {
+                // Sets the site-specific setting
+                if var shieldOverrideEnabled = shieldState.isShieldOverrideEnabled(shield) {
+                    if shield == .AllOff {
+                        // Reversed...
+                        shieldOverrideEnabled = !shieldOverrideEnabled
+                        // TODO: Switch ^ to `isEnabled.toggle()` when we support Swift 4.2+
+                    }
+                    view.toggleSwitch.isOn = shieldOverrideEnabled
+                }
+            }
+        }
+        updateGlobalShieldState(shieldsView.shieldOverrideControl.toggleSwitch.isOn)
+    }
+    
+    private func updateShieldBlockStats() {
+        guard let stats = shieldBlockStats else {
+            // All 0
+            shieldsView.shieldsContainerStackView.adsTrackersStatView.valueLabel.text = "0"
+            shieldsView.shieldsContainerStackView.httpsUpgradesStatView.valueLabel.text = "0"
+            shieldsView.shieldsContainerStackView.scriptsBlockedStatView.valueLabel.text = "0"
+            shieldsView.shieldsContainerStackView.fingerprintingStatView.valueLabel.text = "0"
+            return
+        }
+        shieldsView.shieldsContainerStackView.adsTrackersStatView.valueLabel.text = String(stats.adsAndTrackers)
+        shieldsView.shieldsContainerStackView.httpsUpgradesStatView.valueLabel.text = String(stats.httpsUpgrades)
+        shieldsView.shieldsContainerStackView.scriptsBlockedStatView.valueLabel.text = String(stats.blockedScripts)
+        shieldsView.shieldsContainerStackView.fingerprintingStatView.valueLabel.text = String(stats.fingerprintingProtection)
+    }
+    
+    private func updateBraveShieldState(shield: BraveShieldState.Shield, on: Bool) {
+        guard let domain = url.normalizedHost else { return }
+        if shield == .AllOff {
+            // Technically we set "all off" when the switch is OFF, unlike all the others
```

Comment:
Looks like we have to do a lot of exceptions to handle `.AllOff` state. Do you think it would be worth to remove it from shields state? Just wondering.

Reply:
Yeah this is weird to be honest. I can't really remove it from the shields state because it does need to be saved & persisted along with other site-specific shield settings.

I could change it to be `AllOn` but it gets weird on the shield state side since we have to add things there returning true if its _not_ in the `state` dictionary.

All-around kind of odd, but I think I'd rather have the one-off checks be on the UI side than the data side 😅